Add online installer script

ali-behjati opened the pull request:
This PR adds an online installer scripts which will enable one line installation of tps using:

```
bash -c "$(curl -fsSL https://raw.githubusercontent.com/ioi-2017/tps/master/online-installer/install.sh)"
```

It will also clone the tps repo to `~/.local/share/tps` (or configured `$TPS_LOCAL_REPO`) to be able to use its data (such as `scripts`) in the future.
 
I tested this script in Mac and Linux (via docker) in both bash and zsh.

ali-behjati committed:
Add online installer script

mirjalali commented on online-installer/install.sh:
```diff
@@ -0,0 +1,151 @@
+# $HOME is defined at the time of login, but it could be unset. If it is unset,
+# a tilde by itself (~) will not be expanded to the current user's home directory.
+# POSIX: https://pubs.opengroup.org/onlinepubs/009696899/basedefs/xbd_chap08.html#tag_08_03
+HOME="${HOME:-$(getent passwd $USER 2>/dev/null | cut -d: -f6)}"
```
```suggestion
HOME="${HOME:-$(getent passwd $USER 2> "/dev/null" | cut -d: -f6)}"
```

mirjalali commented on online-installer/install.sh:
```diff
@@ -0,0 +1,151 @@
+BRANCH=${BRANCH:-master}
+
+command_exists() {
+	command -v "$@" >/dev/null 2>&1
```
```suggestion
	command -v "$@" &> "/dev/null"
```

mirjalali commented on online-installer/install.sh:
```diff
@@ -0,0 +1,151 @@
+HOME="${HOME:-$(eval echo ~$USER)}"
+
+# Directory of TPS code repository
+TPS="${TPS:-$HOME/.local/share/tps}"
+
+# Tps cli file path. It will be a symlink from the code repository and the directory should be added to the path.
+TPS_BIN="${TPS_BIN:-$HOME/.local/bin/tps}"
+REPO=${REPO:-ioi-2017/tps}
+REMOTE=${REMOTE:-https://github.com/${REPO}.git}
+BRANCH=${BRANCH:-master}
```
I suggest inlining `REPO` into `REMOTE` as some users might want to use non-github repositories.

Additionally, the name of these environment variables seems to be too general and might be defined for other purposes which may cause problems here.
I recommend names such as `TPS_REMOTE_REPO` and `TPS_REMOTE_BRANCH`.

The name of variable `TPS` is also too general and ambiguous.
Something like `TPS_LOCAL_REPO` seems to be better.

ali-behjati replied:
Makes sense in general.

About your first comment, one can set `REMOTE` entirely and it's easier to keep it as it is right now.

ali-behjati committed:
Update online-installer/install.sh

Co-authored-by: mirjalali <mirjalali@users.noreply.github.com>

ali-behjati committed:
Update online-installer/install.sh

Co-authored-by: mirjalali <mirjalali@users.noreply.github.com>

ali-behjati committed:
Address review comments about naming

ali-behjati committed:
Update script to use install-tps.sh

This wipes out local installation part of the code

ali-behjati committed:
Merge branch 'abehjati/add-autoinstall' of github.com:ali-bahjati/tps into abehjati/add-autoinstall

ali-behjati committed:
Improve comments

ali-behjati committed:
Refactor the code

To make code consistent with other codebase

ali-behjati committed:
Add new installation method to README

ali-behjati committed:
Refactor the code again

ali-behjati committed:
Update Readme and the docs readme

mirjalali commented on online-installer/install.sh:
```diff
@@ -0,0 +1,157 @@
+
+	if user_can_sudo; then
+		echo "You might need to enter your password for installation."
+		sudo -k bash "install-tps.sh"
+	else
+		bash "install-tps.sh"
+	fi 
+
+	# Checking if installation was successful
+	if [ $? -ne 0 ]; then
```
```suggestion
		sudo -k bash "install-tps.sh"
		install_exit_code=$?
	else
		bash "install-tps.sh"
		install_exit_code=$?
	fi 

	# Checking if installation was successful
	if [ "${install_exit_code}" -ne "0" ]; then
```

ali-behjati committed:
Refactor installation script

Co-authored-by: mirjalali <mirjalali@users.noreply.github.com>

ali-behjati committed:
Remove `set -e` from the script

In the script sometimes it handled exit codes itself.
It also handled all exit conditions manually.

ali-behjati committed:
Bring back `set -e`
Now it handles exit usecase of installation in a better way

ali-behjati committed:
Format indentation

ali-behjati committed:
Remove TPS_REMOTE_REPO

ali-behjati committed:
Update doc's Readme

mirjalali commented on online-installer/install.sh:
```diff
@@ -0,0 +1,158 @@
+
+	ostype=$(uname)
+	if [ -z "${ostype%CYGWIN*}" ] && git --version | grep -q msysgit; then
+		fmt_error "Windows/MSYS Git is not supported on Cygwin"
+		fmt_error "Make sure the Cygwin git package is installed and is first on the \$PATH"
```
This code makes `Error: ` be printed twice. Is this intentional?

ali-behjati replied:
Thanks. Fixed.

mirjalali commented on online-installer/install.sh:
```diff
@@ -0,0 +1,158 @@
+		sudo -k bash "install-tps.sh" || install_exit_code=$?
+	else
+		bash "install-tps.sh"
+		install_exit_code=$? || install_exit_code=$?
```
Bug in the latest change

ali-behjati replied:
Thanks. Fixed.

ali-behjati committed:
Address review comments

Fix bugs, ..

ali-behjati committed:
Small refactor

mirjalali commented on online-installer/install.sh:
```diff
@@ -0,0 +1,157 @@
+
+	ostype=$(uname)
+	if [ -z "${ostype%CYGWIN*}" ] && git --version | grep -q msysgit; then
+		fmt_error "Windows/MSYS Git is not supported on Cygwin." \
+			"Make sure the Cygwin git package is installed and is first on the \$PATH"
```
```suggestion
		fmt_error "\
Windows/MSYS Git is not supported on Cygwin.
Make sure the Cygwin git package is installed and is first on the \$PATH."
```

ali-behjati replied:
I think it's better for the error to be in one line.
